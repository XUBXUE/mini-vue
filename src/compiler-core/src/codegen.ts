import { isString } from "../../shared";
import { NodeTypes } from "./ast";
import {
  CREATE_ELEMENT_VNODE,
  helpersMapName,
  TO_DISPLAY_STRING,
} from "./runtimeHelpers";

export function generate(ast) {
  const context = createCodegenContext();
  const { push } = context;

  genFunctionPreamble(ast, context);

  const functionName = "render";
  const args = ["_ctx", "_cache"];
  const signature = args.join(", ");

  push(`function ${functionName}(${signature}) {`);
  push("return ");
  genNode(ast.codegenNode, context);
  push("}");

  return {
    code: context.code,
  };
}

function genFunctionPreamble(ast, context) {
  const { push } = context;
  const vueBinging = "Vue";
  const aliasHelper = (i) => `${helpersMapName[i]}: _${helpersMapName[i]}`;
  if (ast.helpers.length > 0) {
    push(
      `const { ${ast.helpers.map(aliasHelper).join(", ")} } = ${vueBinging}`
    );
  }
  push("\n");
  push("return ");
}

function genNode(node: any, context: any) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context);
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context);
      break;
    case NodeTypes.ELEMENT:
      genElement(node, context);
      break;
    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context);
      break;
    default:
      break;
  }
}

function genText(node: any, context: any) {
  const { push } = context;
  push(`'${node.content}'`);
}

function genInterpolation(node: any, context: any) {
  const { push, helper } = context;
  push(`${helper(TO_DISPLAY_STRING)}(`);
  genNode(node.content, context);
  push(")");
}

function genExpression(node: any, context: any) {
  const { push } = context;
  push(`${node.content}`);
}

function genElement(node: any, context: any) {
  const { push, helper } = context;
  const { tag, children, props } = node; // 由于实现简易版 只有一个符合类型的子节点
  push(`${helper(CREATE_ELEMENT_VNODE)}(`);
  // genNode(children, context);
  genNodeList(genNullable([tag, props, children]), context);
  // const child = children[0]; // 由于实现简易版 只有一个符合类型的子节点 所以直接取0索引
  // for (let i = 0; i < children.length; i++) {
  //   const child = children[i];
  //   genNode(child, context);
  // }
  // genNode(child, context);
  push(")");
}

function genCompoundExpression(node: any, context: any) {
  const { push } = context;
  const { children } = node;
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    if (isString(child)) {
      push(child);
    } else {
      genNode(child, context);
    }
  }
}

function createCodegenContext() {
  const context = {
    code: "",
    push(source) {
      context.code += source;
    },
    helper(key) {
      return `_${helpersMapName[key]}`;
    },
  };
  return context;
}
function genNullable(args: any[]) {
  return args.map((arg) => arg || "null");
}

function genNodeList(args: any[], context) {
  const { push } = context;
  for (let i = 0; i < args.length; i++) {
    const node = args[i];
    if (isString(node)) {
      push(node);
    } else {
      genNode(node, context);
    }
    if (i < args.length - 1) {
      push(", ");
    }
  }
}
