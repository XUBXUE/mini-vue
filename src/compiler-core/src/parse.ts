import { NodeTypes } from "./ast";

export function baseParse(content) {
  const context = createParserContext(content);
  return createRoot(parseChildren(context));
}

function parseChildren(context) {
  const nodes: any[] = [];
  let node;
  if (context.source.startsWith("{{")) {
    node = parseInterpolation(context);
  } else if (context.source[0] == "<") {
    node = parseElement(context);
  }
  nodes.push(node);
  return nodes;
}

function parseElement(context) {
  console.log("context.source", context.source);
  const tagReg = /^<\/?([a-z]*)/i;
  const match: any = tagReg.exec(context.source);
  console.log("match", match);
  const tag = match[1];

  console.log("context.source2", context.source);
  advanceBy(context.source, match[0].length);

  return {
    type: NodeTypes.ELEMENT,
    tag,
  };
}

function parseInterpolation(context) {
  const openDelimiter = "{{";
  const closeDelimiter = "}}";
  // 获取插值结束符号的索引位置
  const closeIndex = context.source.indexOf(
    closeDelimiter,
    openDelimiter.length
  );
  // 将插值开始符号去除（推进操作）
  advanceBy(context, openDelimiter.length);
  // 获取插值表达式的长度
  const rawContentLength = closeIndex - openDelimiter.length;
  // 截取插值表达式
  const rawContent = context.source.slice(0, rawContentLength);
  const content = rawContent.trim();
  // 推进
  advanceBy(context, rawContentLength + closeDelimiter.length);

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: content,
    },
  };
}

function advanceBy(context: any, length) {
  context.source = context.source.slice(length);
}

function createRoot(children) {
  return {
    children,
  };
}

function createParserContext(content: string) {
  return {
    source: content,
  };
}
