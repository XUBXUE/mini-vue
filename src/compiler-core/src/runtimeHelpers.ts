export const TO_DISPLAY_STRING = Symbol("toDisplayString");

export const helpersMapName = {
  [TO_DISPLAY_STRING]: "toDisplayString",
};
