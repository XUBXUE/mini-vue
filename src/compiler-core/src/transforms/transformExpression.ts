import { NodeTypes } from "../ast";

export function transformExpression(node) {
  if (node.type == NodeTypes.INTERPOLATION) {
    node.content = processsExpression(node.content);
  }
}
function processsExpression(node) {
  node.content = `_ctx.${node.content}`;
  return node;
}
