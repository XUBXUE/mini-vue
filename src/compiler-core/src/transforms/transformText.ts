import { NodeTypes } from "../ast";
import { isText } from "../utils";

export function transformText(node) {
  if (node.type == NodeTypes.ELEMENT) {
    return () => {
      const { children } = node;
      let currentContainer;
      for (let i = 0; i < children.length; i++) {
        const currentChild = children[i];
        if (isText(currentChild)) {
          for (let j = i + 1; j < children.length; j++) {
            const nextChild = children[j];
            if (isText(nextChild)) {
              // 如果相邻两个节点为element或text节点，则可以组成复合表达式节点
              // 将当前child节点更改为复合节点 并将本身添加到children里
              if (!currentContainer) {
                currentContainer = children[i] = {
                  type: NodeTypes.COMPOUND_EXPRESSION,
                  children: [currentChild],
                };
              }
              // 将相邻的child节点也添加到children里
              currentContainer.children.push(" + ");
              currentContainer.children.push(nextChild);
              // 去除当前相邻的节点, 并减少索引
              children.splice(j, 1);
              j--;
            } else {
              currentContainer = undefined;
              break;
            }
          }
        }
      }
    };
  }
}
