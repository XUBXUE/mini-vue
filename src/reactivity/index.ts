export { ref } from "./ref.js";
