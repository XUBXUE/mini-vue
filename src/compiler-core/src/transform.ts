import { NodeTypes } from "./ast";
import { TO_DISPLAY_STRING } from "./runtimeHelpers";

export function transform(root, options = {}) {
  const context = createTransfromContext(root, options);
  traverseNode(context.root, context);
  createRootCodegen(root);

  root.helpers = [...context.helpers.keys()];
}

function createRootCodegen(root) {
  // 由于实现的是简易版的compiler，只有一个复合类型子节点 所以取0索引
  const child = root.children[0];
  if (child.type == NodeTypes.ELEMENT) {
    root.codegenNode = child.codegenNode;
  } else {
    root.codegenNode = root.children[0];
  }
}

function traverseNode(node: any, context) {
  const nodeTransforms = context.nodeTransforms;
  const exitFns: any = [];
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i];
    const onExit = transform(node, context);
    if (onExit) exitFns.push(onExit);
  }

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING);
      break;
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traversChildren(node, context);
      break;
    default:
      break;
  }
  let i = exitFns.length;
  while (i--) {
    exitFns[i]();
  }
}

function traversChildren(node, context) {
  const children = node.children;
  if (children) {
    for (let i = 0; i < children.length; i++) {
      const node = children[i];
      traverseNode(node, context);
    }
  }
}

function createTransfromContext(root, options) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, 1);
    },
  };

  return context;
}
