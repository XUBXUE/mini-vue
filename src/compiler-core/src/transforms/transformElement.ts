import { createVNodeCall, NodeTypes } from "../ast";
import { CREATE_ELEMENT_VNODE } from "../runtimeHelpers";

export function transformElement(node, context) {
  if (node.type == NodeTypes.ELEMENT) {
    return () => {
      const vnodeTag = `'${node.tag}'`;

      let vnodeProps;

      const vnodeChildren = node.children[0]; // 由于实现的是简易版的compiler，只有一个复合类型子节点 所以取0索引

      node.codegenNode = createVNodeCall(
        context,
        vnodeTag,
        vnodeProps,
        vnodeChildren
      );
    };
  }
}
