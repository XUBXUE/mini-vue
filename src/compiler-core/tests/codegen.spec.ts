import { generate } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import { transformExpression } from "../src/transforms/transformExpression";

describe("codegen", () => {
  it("text", () => {
    const ast = baseParse("1234");
    transform(ast);
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  it("interpolation", () => {
    const ast = baseParse("{{message}}");
    transform(ast, {
      nodeTransforms: [transformExpression],
    });
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  it("element", () => {
    const ast = baseParse("<div></div>");
    transform(ast, {
      nodeTransforms: [],
    });
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });
});
