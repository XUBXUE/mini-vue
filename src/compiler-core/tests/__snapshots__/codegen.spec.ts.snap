// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
"const { createElementVnode: _createElementVnode } = Vue
return function render(_ctx, _cache) {return _createElementVnode('div', null, 'Hi, ' + _toDisplayString(message)}"
`;

exports[`codegen interpolation 1`] = `
"const { toDisplayString: _toDisplayString } = Vue
return function render(_ctx, _cache) {return _toDisplayString(_ctx.message)}"
`;

exports[`codegen text 1`] = `
"
return function render(_ctx, _cache) {return '1234'}"
`;
